Upstream sessions

jtojnar opened the pull request:
###### Motivation for this change
See https://github.com/NixOS/nixpkgs/issues/39871

This is just a first swallow, in the future pull requests we will want to convert more DEs and investigate Wayland.

###### Things done
- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
- [x] Tested via one or more NixOS tests
     - [x] `gnome3`
     - [x] `gnome3-gdm`
- [x] Tried to run upstream xsession (GNOME) in VM
     - [x] with GDM
     - [x] with lightdm
     - [x] with SDDM
- [x] Tried to run legacy xsession (Xfce) in VM
     - [x] with lightdm
     - [x] with gdm
- [x] Tried to run legacy combined xsession (xmonad+xterm) in VM
     - [x] with lightdm
     - [x] with gdm
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

cc @hedning, @worldofpeace

jtojnar committed:
nixos/services.xserver.displayManager: move X sessions to a subdirectory

Previously, the mkDesktops function produced a flat package containing
session files in the top level. As a preparation for introduction of
Wayland sessions, the files will now be placed to $out/share/xsessions.

jtojnar committed:
nixos/gnome3: rely on xdg autostart for xdg-user-dirs

jtojnar committed:
nixos/display-manager.default: add extraSessionFilePackages option

jtojnar committed:
nixos/gnome3: use upstream session file

Pass gnome-session to extraSessionFilePackages, remove unnecessary environment variables, move the rest out of old session option, and then drop the option.

jtojnar committed:
gnome3.gdm: add upstream xsession

jtojnar committed:
gnome3.gpaste: hard-code paths

GPaste GNOME Shell extension uses GPaste library generated via introspection. Previously, we added the gpaste package to services.xserver.desktopManager.gnome3.sessionPath option, which
added its typelib directory to GI_TYPELIB_PATH environment variable globally, in order for GNOME Shell to be able to find it. This is not very Nix-y, though, so we have decided to patch the code to
append the path to the GI repository search path.

Additionally, the code relies on GPaste’s GSettings schemas, so we had to hard-code the paths to them as well. We ignored the GNOME Shell’s schemas, since they will already be available for the
extension inside GNOME Shell program.

jtojnar committed:
nixos/gnome3: remove session path

Global environment variables are not very Nix-y.

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43992&attempt_id=6ea82175-a073-430a-b1e8-c87f3cedbb9b)

Attempted: gnome3.gdm, gnome3.gpaste

<details><summary>Partial log (click to expand)</summary><p>

```
cannot build derivation '/nix/store/rnnrdjhvrarz8gmvfsmwq5lbpq2914iq-libgdata-0.17.9.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/rp8fzf629m8rx183yy5qy1dn2r530a7x-evolution-data-server-3.28.3.drv': 3 dependencies couldn't be built
cannot build derivation '/nix/store/qf87jknlll6jn3ix547l6nvrzspqf0g6-grilo-plugins-0.3.5.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/6j3fiy4b8b8s865griqk6p3dg71cyn2z-tracker-miners-2.0.5.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/mb30fs25s83y8msliivqj814pbwnxd2m-nautilus-3.28.1.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/4x8933079xainrzq1db39r5mbjacsbbq-totem-3.26.1.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/7awmzq1rhzbw7z59zw7zy3gnj97y4nd9-cheese-3.28.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/lr6574abdh70w9byyx4f7irp1ia8q6f0-gnome-control-center-3.28.2.drv': 4 dependencies couldn't be built
cannot build derivation '/nix/store/sx6qhvgpjcfh3jlr0sinsl7vgjnm9jwa-gpaste-3.28.2.drv': 1 dependencies couldn't be built
error: build of '/nix/store/sx6qhvgpjcfh3jlr0sinsl7vgjnm9jwa-gpaste-3.28.2.drv' failed
```
</p></details>

GrahamcOfBorg commented:
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43992&attempt_id=cf3d4757-7a02-46f2-be2e-f0cbd6ea3c81)

Attempted: gnome3.gdm, gnome3.gpaste

<details><summary>Partial log (click to expand)</summary><p>

```
cannot build derivation '/nix/store/lb3h5kiynhps8dafikvzbrdn46q3af7d-libgdata-0.17.9.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/ryvb61q946n7s6mk2pm0k6yi36a2syij-evolution-data-server-3.28.3.drv': 4 dependencies couldn't be built
cannot build derivation '/nix/store/lgch03spbz8mnsp92nxxhq3xxg9fqz08-grilo-plugins-0.3.5.drv': 3 dependencies couldn't be built
cannot build derivation '/nix/store/ysn1xfa6xqhn12jfyih1n1b6nb59k2aq-tracker-miners-2.0.5.drv': 4 dependencies couldn't be built
cannot build derivation '/nix/store/9nyfnlhrcmjwz5xgl9wzi1fnvf5m0yks-nautilus-3.28.1.drv': 5 dependencies couldn't be built
cannot build derivation '/nix/store/8zplz5hcjw7hnm7bki1brvh283ppj7qd-totem-3.26.1.drv': 7 dependencies couldn't be built
cannot build derivation '/nix/store/9l2n51vhd9syqzn4kfwaap76n04h0lyi-cheese-3.28.0.drv': 5 dependencies couldn't be built
cannot build derivation '/nix/store/k1961fyvc9ihk96z5g84z832vyjv0qpx-gnome-control-center-3.28.2.drv': 11 dependencies couldn't be built
cannot build derivation '/nix/store/9bmz49x5zamxzcidl02nzxish9y7pn5b-gpaste-3.28.2.drv': 4 dependencies couldn't be built
error: build of '/nix/store/3yb1f44wmdpz6ivgcz9rr9lb7y9xjjq2-gdm-3.28.2.drv', '/nix/store/9bmz49x5zamxzcidl02nzxish9y7pn5b-gpaste-3.28.2.drv' failed
```
</p></details>

hedning commented:
This seems like good start :)

Did a little testing (with gdm):

#### Login

1. Tested in a VM without password, which worked as expected
2. Tested on hardware with my regular user and couldn't log in

I'm getting this in the log:
```
juli 23 13:10:16 x1 gdm-password][4245]: gkr-pam: no password is available for user
```

Not sure how to reproduce in a VM as the user.password option is ignored for some reason.

#### GPaste

Gnome-control-center can't find the necessary gpaste schema, so it crashes:

1. Started Gnome Shell in a VM (passwordless user) with `services.gnome3.gpaste.enabled = true`. 
2. Turned on the gpaste extension.
3. Tried launching gnome-control-center

Resulting in:
```
Jul 23 11:27:04 nixos .gnome-control-[1908]: Settings schema 'org.gnome.GPaste' is not installed
```

jtojnar commented:
Add Login, maybe this:

> If the option `users.mutableUsers` is true, the password defined in one of the three options will only be set when the user is created for the first time.

Ad GPaste, that is probably the keyboard shortcuts XML file. I do not think it is possible  Maybe we could create `gnome-control-center-with-plugins` expression. Or just give up and re-add the session path (though I am not sure how to do it).

I tried lightdm now, but it does not even see the session. Also in the log, I see ** (lightdm-gtk-greeter:802): WARNING **: 11:41:49.170: [PIDs] Failed to execute command: /usr/lib/at-spi2-core/at-spi-bus-launcher

hedning commented:
Right, setting mutable users to false made login work with a hardcoded password in a VM, not sure why it won't work on my hardware though (tested on two machines now, will investigate some more).

jtojnar commented:
Regarding GPaste, I guess, we also could extend the `KeyListEntries` in the XML file with `schemadir` attribute and patch Control Center to respect it.

Mic92 commented on nixos/modules/services/x11/desktop-managers/gnome3.nix:
```diff
@@ -55,14 +55,6 @@ in {
         description = "Enable Gnome 3 desktop manager.";
       };
 
-      sessionPath = mkOption {
```
Is there a way to redirect users to the new option (extraSessionFilePackages?) instead?

jtojnar replied:
It serves a completely different purpose. The replacement to `sessionPath` is fixing packages.

hedning commented:
(had a bad `.xprofile` which caused the crash).

Another option for GPaste would be to compile the schema in `/run/current-system/sw/share/glib-2.0/...`, not sure if that's possible in a module though?

jtojnar commented:
Yep, doing as you suggest in https://github.com/NixOS/nixpkgs/issues/42176#issuecomment-399386269 would be another possibility.

worldofpeace commented:
Hmm, sddm doesn't see the session either. Maybe that's missing something.

worldofpeace commented:
This should be fine
```diff
diff --git a/pkgs/desktops/gnome-3/core/gnome-session/default.nix b/pkgs/desktops/gnome-3/core/gnome-session/default.nix
index 7b407aad506..6f9cc384fa2 100644
--- a/pkgs/desktops/gnome-3/core/gnome-session/default.nix
+++ b/pkgs/desktops/gnome-3/core/gnome-session/default.nix
@@ -33,8 +33,10 @@ stdenv.mkDerivation rec {
   preFixup = ''
     for desktopFile in $(grep -rl "Exec=gnome-session" $out/share)
     do
-      echo "Patching gnome-session path in: $desktopFile"
+      echo "Patching Exec gnome-session path in: $desktopFile"
       sed -i "s,^Exec=gnome-session,Exec=$out/bin/gnome-session," $desktopFile
+      echo "Patching TryExec gnome-sesison path in: $desktopFile"
+      sed -i "s,^TryExec=gnome-session,TryExec=$out/bin/gnome-session," $desktopFile
     done
     wrapProgram "$out/bin/gnome-session" \
       --prefix PATH : "${glib.dev}/bin" \

```

jtojnar committed:
gnome3.gnome-session: use absolute paths

TryExec needs absolute path too, otherwise the desktop file will be ignored
unless gnome-session is in PATH, in which case, we would not need to patch
Exec.

worldofpeace commented:
Also,
```nix
debug = mkEnableOption "gnome-session debug messages";
```

It would be nice to still have an interface to do that.

worldofpeace commented:
Do these lightdm [errors](https://gist.github.com/worldofpeace/bec0702389e6f23c8670ded86c7233d2) look of any interest?

jtojnar commented:
Not particularly.

With the latest commit, I am able to launch GNOME Shell in lightdm in VM, but it does not seem to receive environment variables (including `PATH`).

worldofpeace reviewed (commented):
Ahh so that's what it was about.

GrahamcOfBorg commented:
No attempt on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43992&attempt_id=46aa3067-900c-446c-be44-926d48d4cd43)

The following builds were skipped because they don't evaluate on x86_64-darwin: gnome3.gdm, gnome3.gnome-session, gnome3.gpaste

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43992&attempt_id=9f71c89b-8598-4201-9d19-a02490609b13)

Attempted: gnome3.gdm, gnome3.gnome-session, gnome3.gpaste

<details><summary>Partial log (click to expand)</summary><p>

```
strip is /nix/store/fzsyclvxz1p2ia83caniy93v8cwarg0j-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/aw3hz1qsfas7hdnpqr8s27zzjnwlgv8g-cheese-3.28.0-man
checking for references to /build in /nix/store/aw3hz1qsfas7hdnpqr8s27zzjnwlgv8g-cheese-3.28.0-man...
shrinking RPATHs of ELF executables and libraries in /nix/store/bbrqw0s42lcfk6fr2rn213hj11hnl41d-cheese-3.28.0-devdoc
strip is /nix/store/fzsyclvxz1p2ia83caniy93v8cwarg0j-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/bbrqw0s42lcfk6fr2rn213hj11hnl41d-cheese-3.28.0-devdoc
checking for references to /build in /nix/store/bbrqw0s42lcfk6fr2rn213hj11hnl41d-cheese-3.28.0-devdoc...
cannot build derivation '/nix/store/vsysn71qypglwai1slisdjsi6sd80mgd-gnome-control-center-3.28.2.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/4zg3d8a52r9534a6y43di3p8s3afg7rc-gpaste-3.28.2.drv': 1 dependencies couldn't be built
error: build of '/nix/store/4zg3d8a52r9534a6y43di3p8s3afg7rc-gpaste-3.28.2.drv', '/nix/store/8vj9kcg09n2nm17c3n5jwidq8znl9f7z-gnome-session-3.28.1.drv' failed
```
</p></details>

GrahamcOfBorg commented:
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43992&attempt_id=78d1f770-fb54-45fe-b583-1eef3c1285f1)

Attempted: gnome3.gdm, gnome3.gnome-session, gnome3.gpaste

<details><summary>Partial log (click to expand)</summary><p>

```
checking for references to /build in /nix/store/ypphf619vmn6z5p67vriczj2na7blsy1-pipewire-0.1.9-dev...
shrinking RPATHs of ELF executables and libraries in /nix/store/2hb889kzpcyk6f3d6bf3w6dy12whbdvg-pipewire-0.1.9-doc
strip is /nix/store/h0nc84ybvb8g724fwysvcghh1675wjli-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/2hb889kzpcyk6f3d6bf3w6dy12whbdvg-pipewire-0.1.9-doc
checking for references to /build in /nix/store/2hb889kzpcyk6f3d6bf3w6dy12whbdvg-pipewire-0.1.9-doc...
cannot build derivation '/nix/store/4wc99hxdlbh9ylngad9ass3wks2c99nc-mutter-3.28.2.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/2d6yjkl45w8cn0b83zhjjapy58ydxh0m-gnome-shell-3.28.2.drv': 14 dependencies couldn't be built
cannot build derivation '/nix/store/1y93zl13zqz9jipb8b35i6ypaqy05nwx-gpaste-3.28.2.drv': 4 dependencies couldn't be built
cannot build derivation '/nix/store/pw75dadrx3licq6n5p813fng7z2y32v3-gnome-session-3.28.1.drv': 3 dependencies couldn't be built
error: build of '/nix/store/1y93zl13zqz9jipb8b35i6ypaqy05nwx-gpaste-3.28.2.drv', '/nix/store/7myq2453dx5zmyjgf7nkp2zrf3n2m4sy-gdm-3.28.2.drv', '/nix/store/pw75dadrx3licq6n5p813fng7z2y32v3-gnome-session-3.28.1.drv' failed
```
</p></details>

hedning commented:
I took a closer look at how LightDM and NixOS sets up the environment and launches sessions. I'll try to lay out things here for some hopefully shared clarity.

#### The current model

LightDM launches sessions like this ([the same thing goes for sddm](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/sddm.nix#L52)):

1.It extracts the `Exec` command of the selected .desktop session file
2. It passes the command to its `session-wrapper` script which is expected to exec the passed command: https://wiki.ubuntu.com/LightDM#Adding_System_Hooks

NixOS takes control by setting the `session-wrapper` to [`services.xserver.displayManagar.session.script`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/lightdm.nix#L52), ie. the huge [`xsession`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/default.nix#L31) script.

The Exec command is by convention [`${cfg.displayManager.session.script} "${n}"`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/default.nix#L196), where `${n}` is the name of the desktop environment (optionally concatted with `+wm-name`).

So lightdm will normally launch something like this: `/nix/store/*xsession /nix/store/*xsession plasma5`. Which is reduced to the `/nix/store/*xsession plasma5` by [this line of code](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/default.nix#L57): 
```bash
      if [ "''${1:0:1}" = "/" ]; then eval exec "$1" "$2" ; fi
```

The `xsession` script will then extract the desktop environment name from the first argument and run the start script ([relevant code](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/default.nix#L160)).

#### Using upstream sessions

When using the upstream gnome.desktop session the above will simply reduce to executing `/nix/store/*gnome-session*/bin/gnome-session`, skipping `. /etc/profile` etc.

By applying this rather naive fix the gnome shell session is launched with at least $PATH etc. correctly set up (works with sddm too, which I think is an improvement #24298):
```diff
modified   nixos/modules/services/x11/display-managers/default.nix
@@ -32,6 +32,8 @@ let
     ''
       #! ${pkgs.bash}/bin/bash
 
+      . /etc/profile
+
       # Expected parameters:
       #   $1 = <desktop-manager>+<window-manager>
 
@@ -59,7 +61,6 @@ let
       # Now it should be safe to assume that the script was called with the
       # expected parameters.
 
-      . /etc/profile
       cd "$HOME"
 
       # The first argument of this script is the session type.
```

#### Actual solution?

The underlying problem seems to be that we're not using the session's `Exec` command to launch sessions as expected, making upstream sessions difficult. We're instead using [`session.start`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/desktop-managers/plasma5.nix#L43) to do the actual launching through a big case statement in `xsession`.

We could create a proper `session-wrapper` which _only_ sets up the environment and then launches the session's `Exec` command. For legacy purposes the `Exec` command would be a reduced `xsession` script mostly containing a big case statement to select the correct session, ie. starting somewhere [around here](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/default.nix#L149).

This should allow us to handle both legacy and upstream sessions I think.

worldofpeace commented:
@hedning Excellent!
I just need to meditate on it a few times to ensure proper absorption.

> This should allow us to handle both legacy and upstream sessions I think.

That is very much the outcome we want at least for now.

:+1:

hedning committed:
nixos/displayManager: Create a common environment wrapper for all dms

This makes it easier to support a wider variety of .desktop session files. In
particular this makes it possible to use both the «legacy» sessions and upstream
session files.

We separate `xsession` into two parts, `xsessionWrapper` and `xsession`.
`xsessionWrapper` sets up the correct environment and then lauches the session's
Exec command (from the .desktop file), falling back to launching the default
window/desktopManager through the `xsession` script (required by at least some
nixos tests).

`xsession` then _only_ handles launching desktop-managers/window-managers defined
through `services.xserver.desktopManager.session`.

hedning commented:
Okay, I think I have something that could work okay. Tested in a VM that both Plasma5 and Gnome Shell launched correctly in LightDM, SDDM and GDM:

af8b14420d99f8c0d9a1fe1b2b07552d66224770: 
```
nixos/displayManager: Make it possible to use upstream sessions

We separate `xsession` into two parts, `xsessionWrapper` and `xsession`.
`xsessionWrapper` sets up the correct environment and then lauches the session's
Exec command (from the .desktop file).

`xsession` the _only_ handles launching desktop-managers/window-managers defined
through `services.xserver.desktopManager.session`.
```
and 
f6ab354cf0a220a662c4300c87597159e17eebe3:
```
nixos/gdm: Use common xsessionWrapper
```

worldofpeace commented:
I've tested my not so secret project with this too ([elementary](https://elementary.io/)) and it works there as well.

It appears you've done (at least well enough) what you've set out to do and retain legacy support :)

hedning commented:
> It appears you've done (at least well enough) what you've set out to do and retain legacy support :)

Yeah, mostly moved some code around so things should hopefully work the same as before.

On the GPaste/sessionPath issue; We could keep `gnome3.sessionPath` by adding the code to `displayManager.sessionCommands`. Checking against eg. `$XDG_CURRENT_DESKTOP` would keep the current behavior of only applying the `sessionPath` when launching a gnome session too. 

That should keep any potential breakage down, making it easier to land this PR? Does that make sense @jtojnar ?

worldofpeace commented:
>On the GPaste/sessionPath issue; We could keep gnome3.sessionPath by adding the code to displayManager.sessionCommands. Checking against eg. $XDG_CURRENT_DESKTOP would keep the current behavior of only applying the sessionPath when launching a gnome session too.

That seems ok for now yet, we should just probably fix it instead of using `sessionPath`.

I personally still need `sessionPath` as I've ran into issues like with nautilus and gpaste a few times already, and don't actually have the time to fix them canonically.

hedning committed:
nixos/desktopManager: Only trace if the default session isn't found

The default session might be found in `extraSessionFilePackages`, but it's not
viable to detect at evaluation time, so emit a warning.

In LightDM instead of checking `defaultSessionName` against
`displayManager.session.names` we rely on the assertions in
`desktopManager` and `windowMananger` and just check that there's at least one
default set. The second assertion could never actually be triggered.

hedning committed:
nixos/tests/gnome3: Explicitly set default desktopManager

Sessions from `extraSessionFilePackages` isn't picked automatically as the
default session.

hedning commented:
Went through all the relevant tests and found some failures:

### xfce

Found and fixed a problem when using slim and `autoLogin` (which made the nixos xfce tests fail). Apparently the `sessionWrapper` (configured through [`login_cmd` in slim](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/slim.nix#L17) isn't passed any arguments when using autoLogin.

That's simply fixed by launching `xsession` from `xsessionWrapper` if it gets no arguments:

https://github.com/hedning/nixpkgs/commit/a0c090d447fe155b5ef9f9d8757b5f77a4e46a3b

(I squashed https://github.com/hedning/nixpkgs/commit/f6ab354cf0a220a662c4300c87597159e17eebe3)

### gnome3

The gnome3 test launches the `xterm` session since `desktopManager.default` gets set incorrectly. 

We need to register sessions added through  `extraSessionFilePackages` as valid session names when setting `desktopManager.default`. This probably means moving `extraSessionFilePackages` to `xserver.desktopManager` and checking the names there.

LightDM also checks the default session in an unnecessary way, so that needs a change. 

A working prototype (these tests pass: sddm.default sddm.autoLogin slim gnome3-gdm gnome3 plasma5 xfce xmonad i3wm):
- Allow default from extraSessionFilePackages: https://github.com/hedning/nixpkgs/commit/999ed1d6f2fec54d1cf77bdd510c73abac6b564a
- fix lightdm etc.: https://github.com/hedning/nixpkgs/commit/1995f44106a9b570fbee82226b6c94217fe1f33c

hedning commented:
@jtojnar I can open a separate PR for the `xsession` split if that's preferable to cherry picking commits?

hedning committed:
nixos/gnome3: Add back `sessionPath`

Implement through `services.xserver.displayManager.sessionCommands`.

hedning committed:
nixos/gnome3: Add back `debug`

GrahamcOfBorg commented:
No attempt on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43992&attempt_id=0e513949-d46a-4dd9-8138-8e08f9dde8a8)

The following builds were skipped because they don't evaluate on x86_64-darwin: gnome3.gdm, gnome3.gnome-session, gnome3.gpaste

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

jtojnar commented on nixos/modules/services/x11/desktop-managers/default.nix:
```diff
@@ -91,9 +99,15 @@ in
         apply = defaultDM:
+          let
+            extraNames = concatMap (extraPackage: concatMap (d: builtins.match "(.*).desktop" d)
+              (builtins.attrNames (builtins.readDir "${extraPackage}/share/xsessions")))
```
Build during evaluation is frowned upon for performance reasons, we will probably need to switch to a checking in later phase.

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43992&attempt_id=a4b5a43a-64a0-4f3a-96e1-c7237d52cdba)

Attempted: gnome3.gdm, gnome3.gnome-session, gnome3.gpaste

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking /nix/store/l7k4ga5yl7rlz56pq5jqwmcwx4mlgzqm-gpaste-3.28.2/bin/.gpaste-client-wrapped
shrinking /nix/store/l7k4ga5yl7rlz56pq5jqwmcwx4mlgzqm-gpaste-3.28.2/lib/libgpaste.so.11.0.1
gzipping man pages under /nix/store/l7k4ga5yl7rlz56pq5jqwmcwx4mlgzqm-gpaste-3.28.2/share/man/
strip is /nix/store/zrs21zqcchgyabjf4xfimncdq16njizc-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/l7k4ga5yl7rlz56pq5jqwmcwx4mlgzqm-gpaste-3.28.2/lib  /nix/store/l7k4ga5yl7rlz56pq5jqwmcwx4mlgzqm-gpaste-3.28.2/libexec  /nix/store/l7k4ga5yl7rlz56pq5jqwmcwx4mlgzqm-gpaste-3.28.2/bin
patching script interpreter paths in /nix/store/l7k4ga5yl7rlz56pq5jqwmcwx4mlgzqm-gpaste-3.28.2
checking for references to /build in /nix/store/l7k4ga5yl7rlz56pq5jqwmcwx4mlgzqm-gpaste-3.28.2...
/nix/store/4s1gxj7w2sfz6vzrb5vrmammmziq3a7g-gdm-3.28.2
/nix/store/3zbsxhgs04fyymjyxaffg18cpfmlldr4-gnome-session-3.28.1
/nix/store/l7k4ga5yl7rlz56pq5jqwmcwx4mlgzqm-gpaste-3.28.2
```
</p></details>

hedning commented:
Pushed the `throw` to `trace` change, and added back `sessionPath` and `debug` implementations (did some rebasing and rewording): https://github.com/hedning/nixpkgs/commits/upstream-sessions

Feel free to sqaush or fixup any code as you see fit btw.

jtojnar committed:
nixos/gnome3: warn against `sessionPath`

worldofpeace commented on nixos/modules/services/x11/desktop-managers/gnome3.nix:
```diff
@@ -58,8 +58,12 @@ in {
+          Additional list of packages to be added to the session search path.
+          Useful for GNOME Shell extensions or GSettings-conditional autostart.
+
+          Note that this should be a last resort, patching the package is preferred (see GPaste).
```
Indeed :smirk:

benley commented on nixos/modules/services/x11/desktop-managers/gnome3.nix:
```diff
@@ -57,8 +57,12 @@ in {
+          Additional list of packages to be added to the session search path.
+          Useful for GNOME Shell extensions or GSettings-conditional autostart.
+
+          Note that this should be a last resort, patching the package is preferred (see GPaste).
```
Really minor punctuation nit:

"Note that this should be a last resort; patching the package is preferred (see GPaste)."
"Note that this should be a last resort, and patching the package is preferred (see GPaste)."

(either of the above is fine)

jtojnar commented:
Thanks @hedning, now it looks like it is ready to go; all tasks completed and tests passing.

There is a space for more clean-ups, for example, removing pulseaudio from the session wrapper (it is supposed to be socket-activated on modern systems), but we can leave it for future PRs.

Asking maintainers of other DEs for their input: @ttuegel (KDE), @romildo (lxqt), @yegortimoshenko (xfce).

hedning commented:
@worldofpeace @jtojnar We could check that the default session exists when actually building the `desktops` derivation perhaps?
https://github.com/NixOS/nixpkgs/blob/d9d92218ec53129a18740f385a3873973c95ec0d/nixos/modules/services/x11/display-managers/default.nix#L175?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/gnome-control-center-wont-start-gpaste-settings-schema-not-installed/7479/4
